### PR TITLE
[ARM] musttail fixes

### PR DESCRIPTION
**Pull request by kiran-isaac**
Backend:
- Caller and callee arguments no longer have to match, just to take up the same space, as they can be changed before the call
- Allowed tail calls if callee and callee both (or neither) use sret, wheras before it would be dissalowed if either used sret
- Allowed tail calls if byval args are used
- Added debug trace for IsEligibleForTailCallOptimisation

Frontend (clang):
- Do not generate extra alloca if sret is used with musttail, as the space for the sret is allocated already

### CODE REVIEW

**Comment by kiran-isaac** [comment]
@jthackray @ostannard

File: llvm/lib/Target/ARM/ARMISelLowering.cpp [hunk 1]
```diff
@@ -3088,70 +3057,44 @@ bool ARMTargetLowering::IsEligibleForTailCallOptimization(
...
     const uint32_t *CalleePreserved = TRI->getCallPreservedMask(MF, CalleeCC);
-    if (!TRI->regmaskSubsetEqual(CallerPreserved, CalleePreserved))
+    if (!TRI->regmaskSubsetEqual(CallerPreserved, CalleePreserved)) {
+      LLVM_DEBUG(dbgs() << "false (not all registers preserved)\n");
       return false;
+    }
   }
 
-  // If Caller's vararg or byval argument has been split between registers and
+  // If Caller's vararg argument has been split between registers and
```
Comment (efriedma-quic) [inline llvm/lib/Target/ARM/ARMISelLowering.cpp:3075]: What case is this trying to detect?  The varargs convention is exactly the same as the non-varargs convention, as far as I remember...

File: llvm/test/CodeGen/ARM/tail-call-float.ll [hunk 2]
```diff
@@ -1,3 +1,4 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
 ; RUN: llc -mtriple armv7 -target-abi aapcs -float-abi soft -O0 -o - < %s \
 ; RUN:   | FileCheck %s -check-prefix CHECK-SOFT -check-prefix CHECK
```
Comment (efriedma-quic) [inline llvm/test/CodeGen/ARM/tail-call-float.ll:3]: Remove the `-check-prefix CHECK` if the test doesn't need it.

File: llvm/test/CodeGen/ARM/struct_byval.ll [hunk 3]
```diff
@@ -67,33 +327,85 @@ declare i32 @e3(ptr nocapture byval(%struct.LargeStruct) align 16 %in) nounwind
 ; We can't do tail call since address of s is passed to the callee and part of
 ; s is in caller's local frame.
```
Comment (efriedma-quic) [inline llvm/test/CodeGen/ARM/struct_byval.ll:328]: Pretty sure this comment is still correct... we can't tail-call here.  r0 points below the bottom of the stack.

**Review by efriedma-quic — COMMENTED** [review]
Please also add a testcase for something like:
```
struct Large { int x[60]; };
void f(long long a, long long b, Large c, Large d);
void g(long long a, long long b, Large c, Large d) {
  [[clang::musttail]] return f(a,b,d,c);
}
```

File: clang/lib/CodeGen/CGCall.cpp [hunk 4]
```diff
@@ -5085,7 +5085,7 @@ RValue CodeGenFunction::EmitCall(const CGFunctionInfo &CallInfo,
   RawAddress SRetAlloca = RawAddress::invalid();
   llvm::Value *UnusedReturnSizePtr = nullptr;
   if (RetAI.isIndirect() || RetAI.isInAlloca() || RetAI.isCoerceAndExpand()) {
-    if (IsVirtualFunctionPointerThunk && RetAI.isIndirect()) {
+    if ((IsVirtualFunctionPointerThunk || IsMustTail) && RetAI.isIndirect()) {
```
[the remark this answers is not in the document]
Comment (kiran-isaac) [reply clang/lib/CodeGen/CGCall.cpp:5088]: I have created other tests for this. Would you suggest putting those tests, along with this change, in a seperate PR?

[the remark this answers is not in the document]
Comment (efriedma-quic) [reply clang/lib/CodeGen/CGCall.cpp:5088]: Yes, please separate this.

[the remark this answers is not in the document]
Comment (kiran-isaac) [reply clang/lib/CodeGen/CGCall.cpp:5088]: Have now done so: #104795. Have requested you as a reviewer

**Comment by kiran-isaac** [comment]
Accidentally pushed changes to this PR to main. Have reverted. see 1a908c6be3317bbbac73e6a6fc52cabefbdebf7d. I meant to push to my fork. Have now done so. Whoops

**Comment by kiran-isaac** [comment]
Force push to remove revert commit, rather than reverting the revert.

**Comment by kiran-isaac** [comment]
Really made a mess of it here. For anyone trying to understand what on earth happened:
- I pushed a commit to main rather than pushing to the branch on my fork. Lesson learned to always use git in the command line rather than pressing the "sync" button on vscode
- I reverted the commit on main, and then merged from main, reverting the changes
- I then force pushed to drop the revert commit rather than reverting the revert commit
- I realised that I had lost my original changes in the chaos, so I did one more force push to cherry pick the two commits I actually wanted rather.

Apologies for any confusion.

File: llvm/test/CodeGen/ARM/tail-call-float.ll (hunk 2 above)
Comment (kiran-isaac) [reply llvm/test/CodeGen/ARM/tail-call-float.ll]: Have removed

File: llvm/include/llvm/CodeGen/CallingConvLower.h [hunk 5]
```diff
@@ -540,6 +540,8 @@ class CCState {
                });
   }
 
+  void dump() const;
```
[the remark this answers is not in the document]
Comment (kiran-isaac) [reply llvm/include/llvm/CodeGen/CallingConvLower.h:544]: Have added

**Comment by kiran-isaac** [comment]
never let me use git again.

Forgot to revert one of my commits to main, causing merge conflicts.

File: llvm/lib/Target/ARM/ARMISelLowering.cpp (hunk 1 above)
Comment (kiran-isaac) [reply llvm/lib/Target/ARM/ARMISelLowering.cpp]: Yes, I think I was wrong with that change, and the test updates subsequently generated were therefore wrong. I have removed this change for now.

File: llvm/test/CodeGen/ARM/struct_byval.ll (hunk 3 above)
Comment (kiran-isaac) [reply llvm/test/CodeGen/ARM/struct_byval.ll]: I agree, I have removed the change to the test

File: llvm/test/CodeGen/ARM/fp-arg-shuffle.ll [hunk 6]
```diff
@@ -4,6 +4,28 @@
 ; CHECK-NOT: vmov
```
Comment (efriedma-quic) [inline llvm/test/CodeGen/ARM/fp-arg-shuffle.ll:4]: Extra check lines

**Review by efriedma-quic — COMMENTED** [review]
What's the state of byval handling with the current version of the patch?

File: llvm/lib/Target/ARM/ARMISelLowering.cpp [hunk 7]
```diff
@@ -3056,18 +3014,26 @@ bool ARMTargetLowering::IsEligibleForTailCallOptimization(
...
+                      << " (guaranteed tail-call CC)\n");
     return CalleeCC == CallerCC;
+  }
 
   // Also avoid sibcall optimization if either caller or callee uses struct
   // return semantics.
   bool isCalleeStructRet = Outs.empty() ? false : Outs[0].Flags.isSRet();
   bool isCallerStructRet = MF.getFunction().hasStructRetAttr();
-  if (isCalleeStructRet || isCallerStructRet)
+  if (isCalleeStructRet != isCallerStructRet) {
```
Comment (efriedma-quic) [inline llvm/lib/Target/ARM/ARMISelLowering.cpp:3033]: If you're going to allow the case where both the caller and callee are sret, you need to check that the sret points to the same address.

Comment (efriedma-quic) [reply llvm/lib/Target/ARM/ARMISelLowering.cpp:3033]: Nevermind, got it confused with a different ABI.  I don't think we need to impose a restriction on sret for 32-bit Arm.

**Comment by kiran-isaac** [comment]
> What's the state of byval handling with the current version of the patch?

byval arguments to a tail call are allowed as long as they are not split between register and stack, same as before. My change to this check was incorrect

**Comment by kiran-isaac** [comment]
> Please also add a testcase for something like:
> 
> ```
> struct Large { int x[60]; };
> void f(long long a, long long b, Large c, Large d);
> void g(long long a, long long b, Large c, Large d) {
>   [[clang::musttail]] return f(a,b,d,c);
> }
> ```

added this testcase (compiled to ir), it works with my patch and does not without

File: llvm/test/CodeGen/ARM/musttail.ll [hunk 8]
```diff
@@ -0,0 +1,43 @@
+; RUN: llc -mtriple=arm-eabi %s -o - | FileCheck %s
```
Comment (efriedma-quic) [inline llvm/test/CodeGen/ARM/musttail.ll:1]: Please generate checks with update_llc_test_checks.py

File: llvm/test/CodeGen/ARM/musttail.ll [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+  ret void
+}
+
+%struct.Large = type { [60 x i32] }
+
+; Function Attrs: mustprogress noinline optnone
+define dso_local void @large_caller(i64 noundef %0, i64 noundef %1, %struct.Large* noundef byval(%struct.Large) align 4 %2, %struct.Large* noundef byval(%struct.Large) align 4 %3) #0 {
+entry:
+; CHECK: 	b	large_callee
+  musttail call void @large_callee(i64 noundef %0, i64 noundef %1, %struct.Large* noundef byval(%struct.Large) align 4 %2, %struct.Large* noundef byval(%struct.Large) align 4 %3)
```
Comment (efriedma-quic) [inline llvm/test/CodeGen/ARM/musttail.ll:39]: Please add tests for both "%0, %1, %2, %3" and "%0, %1, %3, %2".

**Comment by ostannard** [comment]
Superseded by #109943.

**Pull request closed without merging**